### PR TITLE
Fix TODO in sample.sample_sequences- Avoid 'leaving last token calculation to while loop'

### PR DESCRIPTION
**Pull request by albertwujj**
Hi,

This change runs the initial model step on the full context, by calling the body() function. I added an 'first' parameter defaulting to False to allow this.

### CODE REVIEW

**Comment by bladedsupernova** [comment]
Are you saying the AI was not considering all my context whatttt...huh? When I tested it it was great. What is the new change in English?

**Comment by albertwujj** [comment]
Run the initial prompt through the model all at once instead of the final token separately. It's just a tiny performance improvement.

**Comment by bladedsupernova** [comment]
How did you find it? Are you a collaborator or simply you deeply analyzed the code of GPT-2? I'm interested in paying someone a lot to understand GPT-2...let me know...

File: src/sample.py [hunk 1]
```diff
@@ -41,32 +41,29 @@ def step(hparams, tokens, past=None):
...
     with tf.name_scope('sample_sequence'):
-        # Don't feed the last context token -- leave that to the loop below
-        # TODO: Would be slightly faster if we called step on the entire context,
-        # rather than leaving the last token transformer calculation to the while loop.
-        context_output = step(hparams, context[:, :-1])
-
-        def body(past, prev, output):
-            next_outputs = step(hparams, prev[:, tf.newaxis], past=past)
+        def body(past, prev, output, first=False):
+            next_outputs = step(hparams, prev if first else prev[:, tf.newaxis], past=past)
```
Comment (WuTheFWasThat) [inline src/sample.py]: how about we just always make it just `prev`, and don't do the later squeeze?

File: src/sample.py [hunk 2]
```diff
@@ -41,32 +41,29 @@ def step(hparams, tokens, past=None):
...
             logits = next_outputs['logits'][:, -1, :]  / tf.to_float(temperature)
             logits = top_k_logits(logits, k=top_k)
             samples = tf.multinomial(logits, num_samples=1, output_dtype=tf.int32)
             return [
-                tf.concat([past, next_outputs['presents']], axis=-2),
+                next_outputs['presents'] if first else tf.concat([past, next_outputs['presents']], axis=-2),
```
Comment (WuTheFWasThat) [inline src/sample.py]: i think this should probably be `if past is not None` instead of `if first`

File: src/sample.py [hunk 3]
```diff
@@ -41,32 +41,29 @@ def step(hparams, tokens, past=None):
...
                 tf.squeeze(samples, axis=[1]),
-                tf.concat([output, samples], axis=1),
+                samples if first else tf.concat([output, samples], axis=1),
             ]
 
+        past, prev, output = body(None, context, None, first=True)
```
Comment (WuTheFWasThat) [inline src/sample.py]: i would remove the `first=True` flag, and change output to just be tf.zeros([batch_size, 0]) or something like that

Comment (albertwujj) [reply src/sample.py]: got it, I was considering that. Mine is simpler to read, but yours is kinda like just better.

**Comment by albertwujj** [comment]
In order to set the seed, I made this single [change](https://user-images.githubusercontent.com/21957230/58300062-0b51fe00-7d96-11e9-954b-065ae58d9a12.png) (not sure how Fire works).
However, it seems that this does not result in the same output for the same prompt. This is from the original GPT-2 that I just forked:
image

And this is from my current fork (which has pulled from upstream):
image

Are there more random seeds to be fixed?

**Comment by WuTheFWasThat** [comment]
tensorflow seeds are dependent on the graph :( silly design

**Comment by albertwujj** [comment]
hmm, how do i fix that?

**Comment by WuTheFWasThat** [comment]
I would just test with k=1 or something.  your outputs definitely look buggy though also

**Comment by albertwujj** [comment]
Maybe. not 'buggy' as I think I did exactly what the TODO requested. I'll take a look at the output, and re-check that the TODO request should not change the output tho.

**Comment by albertwujj** [comment]
Remember this is 117M

**Comment by albertwujj** [comment]
Actually, is there anyone you can ask about the TODO request? I do not think it would change the output.

**Comment by albertwujj** [comment]
Sorry, what do you mean by set k to 1?

**Comment by albertwujj** [comment]
To put it very briefly, the TODO does not change the output because all Decoder states from past iterations are passed to the current iteration.

**Comment by albertwujj** [comment]
Close but not quite.
Original:
image

My fork:
image

Same prompt as above. I'll look into it.

**Comment by albertwujj** [comment]
Actually need to double-check the random seeds are fixed

**Comment by albertwujj** [comment]
Wait, about your 'depends on graph' comment. I did not make your suggested change yet, I am testing with this, see my fork for the exact repo I am using.

**Comment by albertwujj** [comment]
Ok yes, it seems like my fork is completely correct except for missing the non-repeating output.
Original:
image
My fork:
image

**Comment by albertwujj** [comment]
Damn the model took the next few words right out of the article: https://www.theverge.com/2019/5/2/18525323/detective-pikachu-review-pokemon (jk obvs article released after)

**Comment by albertwujj** [comment]
Testing one more time with the prompt as 'Hi'

**Comment by albertwujj** [comment]
Juuuuust in case, tested on Unconditional, with same seed fixes at the top of file:
orig
image
my fork
image
Missing a newline!

**Comment by albertwujj** [comment]
Certainly let me know if someone at OpenAI finds out why!

**Comment by albertwujj** [comment]
To summarize the above, I am testing my fork with the TODO fixed, against the actual repo, with the seeds fixed and top_k=1. It seems to match exactly, except for missing the non-repeating output in the beginning.

**Comment by albertwujj** [comment]
Ok, looking at the code again I found the pretty obvious bug. Will fix and test.

**Comment by albertwujj** [comment]
Ok, comparing my code with the original, I think I see where I had a bug. Will fix and test.

**Comment by albertwujj** [comment]
Ok, the commit has fixed the bug. I tested again and got this for both the original and my fork:
image

I wrote down exactly (but in general terms) what past, prev, and output are before any while loop iterations, then what they are after 1 iteration, in the original repo. Then I wrote down what they are after 1 iteration of my method (before and after my latest commit). My latest commit makes them the exact same.

Do you want me to test on any more inputs? I will likely test on the above inputs anyways.

**Comment by albertwujj** [comment]
I will work on removing the 'first' parameter.

File: src/sample.py (hunk 3 above)
Comment (albertwujj) [reply src/sample.py]: Ok, so this was actually the source of the bug. We need to pass the context as the output.

**Comment by albertwujj** [comment]
Ok, tested after removing first param, same result:
image
Bug fixed! Let me know if you want me to test on other inputs.

**Comment by albertwujj** [comment]
slightly different (better) for Detective Pikachu, but I feel this is due to some other source of randomness:
image

**Comment by albertwujj** [comment]
Tested on the 3 prompts, plus Unconditional. Results of this fork are the exact same as the actual repo. 

top_k set to 1 and seeded with 1957, in addition to adding 
'''tf.random.set_random_seed(1957)
np.random.seed(1957)'''
to the top of the script just in case.

**Comment by albertwujj** [comment]
The fork is complete. Barring any further tests you may have. Output matches exactly.

**Comment by albertwujj** [comment]
Just curious, may some of the changes (like shape invariant) cause complications for library users? Or have you just not gotten to looking and checking yet?

**Comment by albertwujj** [comment]
I can write a test script with the two versions of sample.py. Where should I put it? Do you have an exact way you want me to go about it?

**Comment by WuTheFWasThat** [comment]
ah sorry, didn't get around to looking for awhile.  looks good!
